### PR TITLE
Improve robustness of atomics (3.2)

### LINKED ISSUE #46203: Editor crashes when polling adb in EditorExportPlatformAndroid after Mutex modernization
**Godot version:**

3.2.4 rc 8651838268

**OS/device including version:**

Kubuntu 20.10
gcc (Ubuntu 10.2.0-13ubuntu1) 10.2.0

**Issue description:**
Editor crashes when opening project, even empty one.

```
libpthread.so.0!__GI___pthread_mutex_lock(pthread_mutex_t * mutex) (pthread_mutex_lock.c:67)
__gthread_mutex_lock(__gthread_mutex_t * __mutex) (/usr/include/x86_64-linux-gnu/c++/10/bits/gthr-default.h:749)
__gthread_recursive_mutex_lock(__gthread_recursive_mutex_t * __mutex) (/usr/include/x86_64-linux-gnu/c++/10/bits/gthr-default.h:811)
std::recursive_mutex::lock(std::recursive_mutex * const this) (/usr/include/c++/10/mutex:106)
MutexImpl<std::recursive_mutex>::lock(const MutexImpl<std::recursive_mutex> * const this) (./core/os/mutex.h:48)
OS_Unix::execute(OS_Unix * const this, const String & p_path, const List<String, DefaultAllocator> & p_arguments, bool p_blocking, OS::ProcessID * r_child_id, String * r_pipe, int * r_exitcode, bool read_stderr, Mutex * p_pipe_mutex) (./drivers/unix/os_unix.cpp:301)
EditorExportPlatformAndroid::_check_for_changes_poll_thread(void * ud) (./platform/android/export/export.cpp:330)
Thread::callback(Thread * p_self, const Thread::Settings & p_settings, Thread::Callback p_callback, void * p_userdata) (./core/os/thread.cpp:68)
std::__invoke_impl<void, void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*>(void (*&&)(Thread *, const Thread::Settings &, void (*)(void *), void *) __f) (/usr/include/c++/10/bits/invoke.h:60)
std::__invoke<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*>(void (*&&)(Thread *, const Thread::Settings &, void (*)(void *), void *) __fn) (/usr/include/c++/10/bits/invoke.h:95)
std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> >::_M_invoke<0ul, 1ul, 2ul, 3ul, 4ul>(std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> > * const this) (/usr/include/c++/10/thread:264)
std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> >::operator()(std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> > * const this) (/usr/include/c++/10/thread:271)
std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> > >::_M_run(std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> > > * const this) (/usr/include/c++/10/thread:215)
execute_native_thread_routine (Unknown Source:0)
libpthread.so.0!start_thread(void * arg) (pthread_create.c:463)
libc.so.6!clone() (clone.S:95)
```

and I got `pthread_mutex_lock.c` said below
```
Could not load source './nptl/../nptl/pthread_mutex_lock.c': 'SourceRequest' not supported..
```

**Steps to reproduce:**
1. Run editor
2. Creating a project
3. Open & edit

**Minimal reproduction project:**

### PR DESCRIPTION
**Pull request by RandomShaper**
Version of #46197 for 3.2.

*Bugsquad edit:* Fixes #46203.

---
**This code is generously donated by IMVU.**

### CODE REVIEW

**Review by hpvb — APPROVED** [review]
Tested with 3.2 tps-demo and some other stuff. LGTM

File: core/safe_refcount.cpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+/*************************************************************************/
+
+#if defined(DEBUG_ENABLED) && !defined(NO_THREADS)
+
+#include "safe_refcount.h"
+
+#include "core/error_macros.h"
+
+// On C++14 we don't have std::atomic::is_always_lockless, so this is the best we can do
+void check_lockless_atomics() {
```
Comment (hpvb) [inline core/safe_refcount.cpp:38]: @akien-mga If you don't like this new warning for release_debug and debug builds the worst case is that it's no worse than it was before we migrated to the new atomic types. Gnu's extensions are also not guaranteed to be lock-less.

**Comment by RandomShaper** [comment]
Added commit to fix #46203, as being also related to the whole MT modernization effort.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**